common/worker_versioning/worker_versioning.go

```diff
@@ -108,12 +108,12 @@ func StampIfUsingVersioning(stamp *commonpb.WorkerVersionStamp) *commonpb.Worker
 	return nil
 }
 
-func MakeDirectiveForWorkflowTask(assignedBuildId string, stamp *commonpb.WorkerVersionStamp, lastWorkflowTaskStartedEventID int64) *taskqueuespb.TaskVersionDirective {
+func MakeDirectiveForWorkflowTask(inheritedBuildId string, assignedBuildId string, stamp *commonpb.WorkerVersionStamp, lastWorkflowTaskStartedEventID int64) *taskqueuespb.TaskVersionDirective {
 	if id := StampIfUsingVersioning(stamp).GetBuildId(); id != "" && assignedBuildId == "" {
 		// TODO: old versioning only [cleanup-old-wv]
 		return MakeBuildIdDirective(id)
-	} else if lastWorkflowTaskStartedEventID == common.EmptyEventID {
-		// first workflow task
+	} else if lastWorkflowTaskStartedEventID == common.EmptyEventID && inheritedBuildId == "" {
```

Comment:
shouldn't there be a branch here where we use inheritedBuildId?

Reply:
No assignedBuildId is set to inheritedBuildId at the very beginning ApplyWorkflowExecutionStartedEvent. Here we only check it so we do not allow using assignment rules for the first WFT if the WF has inherited a build id. Updated the comment to be more clear

Reply:
In that case, can't we let the inherited build id take precedence over "use assignment rules" by just switching the order of these two clauses? if assignedBuildId != "" then use that, else if is-first-wft then use assignment rules?

Reply:
no, assignedBuildId can be populated from the last attempt. we want the second attempt of the first wft use  the latest assignment rules. despite the name, `lastWorkflowTaskStartedEventID` is the last started event id of the last completed task, so it can be empty but it doesn't mean it's the first wft being started.

Reply:
ah. I knew the semantics of `lastWorkflowTaskStartedEventID` but missed the fact that `assignedBuildId` can be populated from a failed attempt. that's kind of subtle.

it makes me want to keep a "tentative assigned id" during the attempt, and "actual assigned id" after the task is completed. but that's probably more complicated than what we have.